File: 2019/Python/day9.py
### =============================================
### [ ADVENT OF CODE ] (https://adventofcode.com)
### 2019 - Mina Pêcheux: Python version
### ---------------------------------------------
### Day 9: Sensor Boost
### =============================================
from intcode import IntcodeProgram

# [ Input parsing functions ]
# ---------------------------
def parse_input(data):
    '''Parses the incoming data into processable inputs.
    
    :param data: Provided problem data.
    :type data: str
    :return: Parsed data.
    :rtype: list(int)
    '''
    return [ int(x) for x in data.split(',') if x != '' ]

# [ Computation functions ]
# -------------------------
### Part I + II
def process_inputs(inputs, input=None, debug=False):
    '''Executes the Intcode program on the provided inputs and computes the final
    result.
    
    :param inputs: List of integers to execute as an Intcode program.
    :type inputs: list(int)
    :param input: Integer to provide as input to the program.
    :type input: int
    :param debug: Whether or not the IntcodeProgram should debug its
        execution at each instruction processing.
    :type debug: bool
    :return: Last output of the program.
    :rtype: int
    '''
    # create program
    program = IntcodeProgram(inputs, debug=debug)
    # insert input in memory if need be
    if input is not None:
        program.push_memory(input)
    # run program
    program.run()
    # get final result
    return program.output[-1]

# [ Base tests ]
# --------------
def make_tests():
    '''Performs tests on the provided examples to check the result of the
    computation functions is ok.'''
    # test new instructions
    ref = '109,1,204,-1,1001,100,1,100,1008,100,16,101,1006,101,0,99'
    program = IntcodeProgram([
        109,1,204,-1,1001,100,1,100,1008,100,16,101,1006,101,0,99
    ])
    program.run()
    assert ','.join([ str(x) for x in program.output ]) == ref
    
    # Part I + II
    assert len(str(process_inputs([ 1102,34915192,34915192,7,4,7,99,0 ]))) == 16
    assert process_inputs([ 104,1125899906842624,99 ]) == 1125899906842624

if __name__ == '__main__':
    # check function results on example cases
    make_tests()
    
    # get input data
    data_path = '../data/day9.txt'
    inputs = parse_input(open(data_path, 'r').read())
    
    ### PART I
    solution = process_inputs(inputs, input=1)
    print('PART I: solution = {}'.format(solution))
    
    ### PART II
    solution = process_inputs(inputs, input=2)
    print('PART II: solution = {}'.format(solution))
    

File: 2019/Javascript/day9.js
/*
 * ================================================
 * [ ADVENT OF CODE ] (https://adventofcode.com)
 * 2019 - Mina Pêcheux: Javascript (NodeJS) version
 * ------------------------------------------------
 * Day 9: Sensor Boost
 * ================================================
 */
'use strict'

const _ = require('lodash')
const fs = require('fs')
const expect = require('chai').expect
const { IntcodeProgram } = require('./utils/intcode')

// [ Input parsing functions ]
// ---------------------------
/**
 * Parses the incoming data into processable inputs.
 * @param {string} data - Provided problem data.
 * @returns {array(int)} - Parsed data.
 */
const parseInput = (data) => {
  return _.map(_.filter(_.split(data, ','), (l) => l.length > 0), Number)
}

// [ Computation functions ]
// -------------------------
/* PART I + II */
/**
 * Executes the Intcode program on the provided inputs and computes the final
   result.
 * @param {array(int)} inputs - List of integers to execute as an Intcode program.
 * @param {int} input - Integer to provide as input to the program.
 * @param {bool} debug - Whether or not the IntcodeProgram should debug its
 *                       execution at each instruction processing.
 * @returns {int} - Last output of the program.
 */
const processInputs = (inputs, input=null, debug=false) => {
  const program = new IntcodeProgram(inputs, debug)
  if (input !== null) {
    program.pushMemory(input)
  }
  program.run()
  return _.last(program.output)
}

// [ Base tests ]
// --------------
/**
 * Performs tests on the provided examples to check the result of the
 * computation functions is ok.
 */
const makeTests = () => {
  // test new instructions
  const ref = '109,1,204,-1,1001,100,1,100,1008,100,16,101,1006,101,0,99'
  const program = new IntcodeProgram([
    109,1,204,-1,1001,100,1,100,1008,100,16,101,1006,101,0,99
  ])
  program.run()
  expect(_.join(program.output, ',')).to.equal(ref)
  
  /* Part I + II */
  const r = processInputs([ 1102,34915192,34915192,7,4,7,99,0 ])
  expect(r.toString().length).to.equal(16)
  expect(processInputs([ 104,1125899906842624,99 ])).to.equal(1125899906842624)
}

(() => {
  // check function results on example cases
  makeTests()
  
  // get input data
  const data = fs.readFileSync('../data/day9.txt')
  const inputs = parseInput(data)
  
  // Part I
  const solution1 = processInputs(inputs, 1)
  console.log(`PART I: solution = ${solution1}`)
  
  // Part II
  const solution2 = processInputs(inputs, 2)
  console.log(`PART II: solution = ${solution2}`)
})()
